commit b5662617959ef558e1130a250a88f9f189cb1bae
Author: Oleg Nesterov <oleg@redhat.com>
Date:   Fri Apr 2 18:05:12 2010 +0200

    tty: release_one_tty() forgets to put pids
    
    release_one_tty(tty) can be called when tty still has a reference
    to pgrp/session. In this case we leak the pid.
    
    Signed-off-by: Oleg Nesterov <oleg@redhat.com>
    Reported-by: Catalin Marinas <catalin.marinas@arm.com>
    Reported-and-tested-by: Tetsuo Handa <penguin-kernel@I-love.SAKURA.ne.jp>
    Acked-by: Linus Torvalds <torvalds@linux-foundation.org>
    Acked-by: Eric W. Biederman <ebiederm@xmission.com>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

diff --git a/drivers/char/tty_io.c b/drivers/char/tty_io.c
index a51374e..60b691e 100644
--- a/drivers/char/tty_io.c
+++ b/drivers/char/tty_io.c
@@ -2342,6 +2342,8 @@ static void release_one_tty(struct tty_struct *tty, int idx)
 	list_del_init(&tty->tty_files);
 	file_list_unlock();
 
+	put_pid(tty->pgrp);
+	put_pid(tty->session);
 	free_tty_struct(tty);
 }
 
